### PR TITLE
feat: Introduce v1alpha2 version of LlamaStackDistribution CRD

### PR DESCRIPTION
**Pull request by VaishnaviHire**
This PR introduces the `v1alpha2` API version for the `LlamaStackDistribution` CRD, enabling declarative, Kubernetes-native configuration of LlamaStack servers. Instead of requiring users to manually craft and supply a `config.yaml` via ConfigMap (as in v1alpha1), the operator now generates the server configuration automatically from structured CR fields (`providers`, `resources`, `storage`, `networking`). Both API versions are served concurrently with full conversion webhook support.

## v1alpha2 Example

The v1alpha2 API replaces environment-variable-driven configuration with structured, declarative fields. All provider fields use typed `[]ProviderConfig` slices with CEL validation.

**Basic : single Ollama provider:**

```yaml
apiVersion: llamastack.io/v1alpha2
kind: LlamaStackDistribution
metadata:
  name: llamastackdistribution-v1alpha2-sample
spec:
  distribution:
    name: starter
  providers:
    inference:
      - provider: ollama
        endpoint: http://ollama-server-service.ollama-dist.svc.cluster.local:11434/v1
  resources:
    models:
      - name: "llama3.2:1b"
  networking:
    port: 8321
  workload:
    replicas: 1
```

**Advanced : vLLM with secret refs, PostgreSQL storage, pgvector:**

```yaml
apiVersion: llamastack.io/v1alpha2
kind: LlamaStackDistribution
metadata:
  name: llamastack-vllm-pg
spec:
  distribution:
    name: starter
  providers:
    inference:
      - provider: vllm
        endpoint: http://vllm-service.vllm.svc.cluster.local:8000/v1
        secretRefs:
          api_key:
            name: vllm-creds
            key: token
    vectorIo:
      - provider: pgvector
        secretRefs:
          host:
            name: pg-credentials
            key: host
        settings:
          port: 5432
          db: llamastack
  resources:
    models:
      - name: llama3.2-8b
  storage:
    kv:
      type: redis
      endpoint: redis://redis-service.redis.svc.cluster.local:6379
    sql:
      type: postgres
      connectionString:
        name: pg-credentials
        key: dsn
  disabled:
    - safety
  workload:
    replicas: 2
```

## Review Guide

This is a large PR (86 files, ~18k lines). The sections below group changes by area matching the commit structure. Each section is self-contained and can be reviewed independently.

### 1. v1alpha2 CRD Schema & Conversion (commit: `e7cc5c2`)

| File | What to review |
|------|---------------|
| `api/v1alpha2/llamastackdistribution_types.go` | New spec/status types. Key design: typed `[]ProviderConfig` slices with CEL validation for provider ID uniqueness. `OverrideConfig` is mutually exclusive with providers/resources/storage/disabled. |
| `api/v1alpha2/zz_generated.deepcopy.go` | Auto-generated |
| `api/v1alpha1/llamastackdistribution_conversion.go` | Bidirectional v1alpha1 ↔ v1alpha2 conversion. Uses JSON blob annotations (`annV1Alpha1Extras`, `annV1Alpha2Extras`) for lossless round-trips in both directions. |
| `api/v1alpha1/llamastackdistribution_conversion_test.go` | Round-trip tests: providers, resources, storage, disabled, TLS, expose hostname, status fields |
| `config/crd/bases/llamastack.io_llamastackdistributions.yaml` | Generated CRD YAML with OpenAPI and CEL rules |

### 2. Validating Webhook (commit: `a45c9f6`)

| File | What to review |
|------|---------------|
| `api/v1alpha2/llamastackdistribution_webhook.go` | Validating webhook: provider ID uniqueness across all API types, distribution name validation, model provider reference checks |
| `api/v1alpha2/llamastackdistribution_webhook_test.go` | Unit tests: cross-slice collision detection, `deriveProviderID` behavior, unknown distribution rejection, edge cases |
| `config/webhook/*` | Webhook service, manifests, kustomize config |
| `config/certmanager/*` | Certificate and issuer for vanilla Kubernetes |
| `config/crd/kustomization.yaml` | Enabled webhook/cert-manager patches |
| `config/default/kustomization.yaml` | Enabled webhook and cert-manager components |
| `config/default/manager_webhook_patch.yaml` | Webhook port and TLS cert volume mount |
| `main.go` | Webhook registration |

### 3. Config Generation Pipeline (commit: `3feb572`)

| File | What to review |
|------|---------------|
| `pkg/config/config.go` | Pipeline: resolve base config → expand providers → expand resources → apply storage → apply disabled APIs → clean registered_resources → override port → render YAML. Key: deep-copy safety, deterministic output |
| `pkg/config/provider.go` | Provider expansion: `remote::` prefix, endpoint → `base_url`, sorted secret ref iteration, settings merge with override protection |
| `pkg/config/resource.go` | Model/tool/shield expansion with default provider resolution and provider existence validation |
| `pkg/config/storage.go` | KV (sqlite/redis) and SQL (sqlite/postgres) with secret env var mapping |
| `pkg/config/secret_resolver.go` | Resolves `secretRefs` maps to env vars (`LLSD_<PROVIDER_ID>_<KEY>`) and `${env.VAR_NAME}` substitutions |
| `pkg/config/resolver.go` | Base config resolver: resolves embedded configs by distribution name |
| `pkg/config/version.go` | Config version detection (supports versions 1-2) |
| `pkg/config/types.go` | Shared types: `BaseConfig`, `ProviderEntry`, `GeneratedConfig` |
| `pkg/config/config_test.go` | Unit tests: determinism, provider/resource expansion, storage, secret resolution, disabled API cleanup, deep-copy safety |
| `pkg/config/configs/*/config.yaml` | Embedded base configs for `starter`, `starter-gpu`, `postgres-demo` distributions |
| `distributions.json` | Distribution metadata |

### 4. Controller Integration (commit: `bf90527`)

| File | What to review |
|------|---------------|
| `controllers/v1alpha2_config.go` | v1alpha2 config handling: determines config source (override / generated / default), creates immutable ConfigMaps with content-hash naming, validates secret/ConfigMap refs, injects secret-backed env vars into pod spec, cleans up old ConfigMaps |
| `controllers/llamastackdistribution_controller.go` | Integration: calls `handleV1Alpha2NativeConfig` before standard reconcile. Dual status update path for v1alpha2 CRs |
| `controllers/kubebuilder_rbac.go` | RBAC markers: added `secrets` (get/list/watch) and `configmaps` (delete) |
| `controllers/resource_helper.go` | Deprecated `startupScript`. Sets `RUN_CONFIG_PATH` env var; image's built-in `entrypoint.sh` handles startup |
| `controllers/resource_helper_test.go` | Updated assertions: `RUN_CONFIG_PATH` instead of command/args overrides |
| `controllers/suite_test.go` | Envtest setup with webhook server |
| `controllers/testing_support_test.go` | Test constants and helpers |
| `controllers/llamastackdistribution_controller_test.go` | Envtest integration tests: config generation, ConfigMap creation, secret env var injection, status updates |
| `config/rbac/role.yaml` | Generated `ClusterRole` with secrets and configmaps-delete permissions |

### 5. OpenShift Webhook Overlay (commit: `e3f405b`)

| File | What to review |
|------|---------------|
| `config/openshift/kustomization.yaml` | OpenShift overlay: replaces cert-manager with service-serving certificates |
| `config/openshift/crd_ca_patch.yaml` | CRD CA injection annotation |
| `config/openshift/manager_webhook_patch.yaml` | Manager cert volume mount |
| `config/openshift/webhook_ca_patch.yaml` | Webhook CA injection annotation |

### 6. E2E Tests (commit: `c6d24e8`)

| File | What to review |
|------|---------------|
| `tests/e2e/creation_v1alpha2_test.go` | v1alpha2 CR creation, ConfigMap generation, Ready phase, secret env var injection into Deployment |
| `tests/e2e/conversion_test.go` | Cross-version read (v1alpha1 as v1alpha2 and vice versa) |
| `tests/e2e/webhook_validation_test.go` | Webhook rejects: missing distribution, duplicate provider IDs, invalid provider references |
| `tests/e2e/validation_test.go` | CRD structure, webhook service/TLS, operator readiness |
| `tests/e2e/creation_test.go` | Updated v1alpha1 creation tests |
| `tests/e2e/e2e_test.go` | Test suite registration |
| `tests/e2e/test_utils.go` | Test helpers |
| `.github/workflows/run-e2e-test.yml` | CI workflow updates for v1alpha2 targets |

### 7. Documentation & Samples (commit: `3b6972b`)

| File | What to review |
|------|---------------|
| `docs/migration-v1alpha1-to-v1alpha2.md` | Migration guide: field mapping tables, before/after examples, step-by-step migration |
| `docs/api-overview.md` | Full v1alpha2 API reference for both versions |
| `README.md` | Updated quick start with v1alpha2 examples using `secretRefs` and list syntax |
| `config/samples/v1alpha1/*` | Existing samples moved into versioned subdirectory |
| `config/samples/v1alpha2/*` | New v1alpha2 samples: basic, HA, vLLM+Postgres, networking |
| `specs/002-operator-generated-config/*` | Updated spec contracts and data model |

### 8. Build Tooling & Release (commit: `b5a2df7`)

| File | What to review |
|------|---------------|
| `Makefile` | Build target updates for v1alpha2 |
| `.gitignore` | Ignore patterns for generated artifacts |
| `go.mod` / `go.sum` | Dependency updates |
| `release/operator.yaml` | Regenerated release manifest with all v1alpha2 resources |

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+		config.ToolGroups = tools
+	}
+	if shields != nil {
+		config.Shields = shields
+	}
+	return resourceCount, nil
+}
+
+// applyNetworkingOverrides applies disabled APIs and port overrides from the CR spec.
+func applyNetworkingOverrides(spec *v1alpha2.LlamaStackDistributionSpec, config *BaseConfig) {
```
Comment (mfleader) [inline pkg/config/config.go]: Missing test coverage for `applyNetworkingOverrides` and `overrideServerPort`.

File: controllers/v1alpha2_config.go [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+type matchingLabelsOption map[string]string
+
+func (o matchingLabelsOption) ApplyToList(opts *client.ListOptions) {
+	sel := labels.SelectorFromSet(map[string]string(o))
+	opts.LabelSelector = sel
+}
+
+// ValidateV1Alpha2SecretRefs validates that all secretKeyRef references in a v1alpha2 spec
+// point to existing Secrets.
+func (r *LlamaStackDistributionReconciler) ValidateV1Alpha2SecretRefs(
```
Comment (mfleader) [inline controllers/v1alpha2_config.go]: I think this should be called in the reconcile path, and covered by a test.

**Review by mfleader — CHANGES_REQUESTED** [review]

File: api/v1alpha2/llamastackdistribution_webhook.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+	}
+
+	if len(allErrs) > 0 {
+		return warnings, allErrs.ToAggregate()
+	}
+
+	return warnings, nil
+}
+
+func validateProviderIDUniqueness(spec *ProvidersSpec) field.ErrorList {
```
Comment (mfleader) [inline api/v1alpha2/llamastackdistribution_webhook.go]: FR-071 and FR-072 say CEL validation, but CEL can't inspect `apiextensionsv1.JSON` fields. I think the webhook is correct here. The spec should be updated to match.

**Review by mfleader — CHANGES_REQUESTED** [review]

File: tests/e2e/webhook_validation_test.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+			},
+			Resources: &v1alpha2.ResourcesSpec{
+				Models: []apiextensionsv1.JSON{{Raw: modelJSON}},
+			},
+		},
+	}
+
+	err := TestEnv.Client.Create(TestEnv.Ctx, cr)
+	require.Error(t, err, "CR with model referencing nonexistent provider should be rejected")
+	if err != nil {
```
Comment (mfleader) [inline tests/e2e/webhook_validation_test.go]: I think this should be `if err == nil` so the CR gets cleaned up when the webhook unexpectedly accepts.

**Review by mfleader — CHANGES_REQUESTED** [review]

File: api/v1alpha2/llamastackdistribution_types.go [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+	Endpoint string `json:"endpoint,omitempty"`
+	// ApiKey references a Secret containing the API key for authentication.
+	// +optional
+	ApiKey *SecretKeyRef `json:"apiKey,omitempty"`
+	// Settings contains provider-specific settings merged into the provider's
+	// config section in config.yaml. Acts as an escape hatch for fields not
+	// directly exposed in the CRD schema. The operator recognizes secretKeyRef
+	// objects only at the top level of settings values.
+	// +optional
+	Settings *apiextensionsv1.JSON `json:"settings,omitempty"`
```
Comment (eoinfennessy) [inline api/v1alpha2/llamastackdistribution_types.go]: Instead of parsing the settings for secret refs, could we have a `SecretRefs` field in the `ProviderConfig`?

**Review by eoinfennessy — COMMENTED** [review]
Added review focussing on API and validation webhooks for now -- generally looks great! I will come back to controller code

**Review by mfleader — CHANGES_REQUESTED** [review]

**Review by eoinfennessy — COMMENTED** [review]
Regarding the API, I think we should copy the stricter validations found here:

raw: https://raw.githubusercontent.com/rhuss/llama-stack-k8s-operator/62acaf509f8d4388b0ff4a02bd084e81532c6ac3/api/v1alpha2/llamastackdistribution_types.go

diff: https://github.com/llamastack/llama-stack-k8s-operator/commit/62acaf509f8d4388b0ff4a02bd084e81532c6ac3

These include (but are not limited to):

- Add MinItems=1 on all optional slices to reject explicitly empty lists
  (provider slices, models, tools, shields, env, command, args, volumes,
  volumeMounts, topologySpreadConstraints, configMapKeys, namespaces, labels)
- Add MinProperties=1 on secretRefs map
- Add MinLength=1 on required string fields (provider, name, configMapName,
  providerId, image, secretKeyRef name/key)
- Add CEL non-empty guards on optional strings (id, provider, modelType,
  quantization, hostname, serviceAccountName, mountPath, configMapNamespace)
- Add reverse validation on storage types: endpoint/password only valid for
  redis, connectionString only valid for postgres
- Add TLS reverse validation: secretName and caBundle only valid when enabled
- Add port range validation (1-65535)
- Add replicas Minimum=0
- Add autoscaling: minReplicas/maxReplicas Minimum=1, percentage range 1-100,
  CEL rule maxReplicas >= minReplicas
- Add PDB mutual exclusivity: minAvailable and maxUnavailable cannot both be
  set, at least one required

In addition, we should...
1. add enum validation for the slice of disabled APIs: `// +kubebuilder:validation:Items:Enum=agents;datasetio;eval;inference;safety;scoring;telemetry;tool_runtime;vector_io`
2. Add CEL validation (or webhook validation if CEL is not possible) to ensure each ID in each `ProviderConfig` slice is unique

**Review by eoinfennessy — COMMENTED** [review]
Final few comments on the API.

File: api/v1alpha2/llamastackdistribution_types.go [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+	// Enabled enables TLS on the server.
+	// +optional
+	Enabled bool `json:"enabled,omitempty"`
+	// SecretName references a Kubernetes TLS Secret. Required when enabled is true.
+	// +optional
+	SecretName string `json:"secretName,omitempty"`
+	// CABundle configures custom CA certificates via ConfigMap reference.
+	// +optional
+	CABundle *CABundleConfig `json:"caBundle,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (VaishnaviHire) [reply api/v1alpha2/llamastackdistribution_types.go]: We don't need the SecretName here. I will remove it from spec. The only required files is ca-bundle configmap

**Review by eoinfennessy — COMMENTED** [review]
Review of config gen pipeline. There are some critical issues that need to be addressed.

**Review by eoinfennessy — COMMENTED** [review]
Review of webhooks:

We need to comprehensively test all validation logic (webhook and CEL). Currently we don't do any testing of validation logic.

There are some problems with data loss and stale data in the conversion logic. I added a suggestion to fix this.

**Review by eoinfennessy — COMMENTED** [review]
I re-reviewed after the previous suggestions. Thanks for addressing these. Couple of things left from these reviews:

1. CEL validation tests: Let's add envtest tests to ensure all of our complex CEL validation is actually working.
2. Split TLSSpec for server and client: see my latest comment in the thread above discussing this
3. Small bug remaining in resource.go (see below)

**Review by eoinfennessy — COMMENTED** [review]
Review focussing on `/controller`. Mostly minor suggestions, but a couple of major things related to surfacing errors and status.

**Comment by VaishnaviHire** [comment]
Closing this for https://github.com/ogx-ai/ogx-k8s-operator/pull/288
